Start adding a "tags" package.

sebright opened the pull request:
This commit adds TagKey and TagSet.  The classes are not used yet.

_______________________________
This change is part of #251.

/cc @dinooliva @songy23 @bogdandrutu

bogdandrutu commented on core/src/main/java/com/google/instrumentation/tags/TagKey.java:
```diff
@@ -0,0 +1,88 @@
+  /** The maximum length for a tag key name. */
+  public static final int MAX_LENGTH = StringUtil.MAX_LENGTH;
+
+  enum TagType {
```
Isn't the type encoded into the generic type?

bogdandrutu commented on core/src/main/java/com/google/instrumentation/tags/TagKey.java:
```diff
@@ -0,0 +1,88 @@
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * Tag key.
```
Please explain more about where is used and how is used. The current comment does not add any value.

sebright replied:
I wasn't sure what to write, but I tried to add an explanation in the TagSet Javadoc.  Are the new comments accurate?

bogdandrutu commented on core/src/main/java/com/google/instrumentation/tags/TagSet.java:
```diff
@@ -0,0 +1,177 @@
+import com.google.instrumentation.internal.StringUtil;
+import javax.annotation.Nullable;
+
+/** A set of tags. */
```
What is a tag?

bogdandrutu commented on core/src/main/java/com/google/instrumentation/tags/TagSet.java:
```diff
@@ -0,0 +1,177 @@
+  // Should these methods take defaults or throw exceptions to handle missing keys?
+
+  /**
+   * Gets a value of type {@code String}.
```
s/Gets/Returns

Maybe change everywhere in this commit?

bogdandrutu commented on core/src/main/java/com/google/instrumentation/tags/TagSet.java:
```diff
@@ -0,0 +1,177 @@
+     * @param value the value to insert for the given key.
+     * @return this
+     */
+    public abstract Builder insert(TagKey<Long> key, long value);
```
We should follow java map or set (probably map) standards and use put or add instead.

adriancole replied:
some of these imply the caller remembering if something was there before or a potentially racy check. Seems laborious vs put semantics. If this labor is worthwhile, we should probably explain in the type level docs why it is.

sebright replied:
> some of these imply the caller remembering if something was there before or a potentially racy check.

What is wrong with a racy check?  It is done in the builder, which is already not thread-safe.

> Seems laborious vs put semantics.  If this labor is worthwhile, we should probably explain in the type level docs why it is.

"set" has the same behavior as "put".  I think that the others are just convenience methods to save the user from having to call tagKeyExists() when they want to do something that depends on the existing value of a key.

/cc @a-veitch

sebright commented:
I have a few questions about the API, which are in TODOs in TagSet.

a-veitch commented on core/src/main/java/com/google/instrumentation/tags/TagKey.java:
```diff
@@ -0,0 +1,89 @@
+@AutoValue
+public abstract class TagKey<TagValueT> {
+  /** The maximum length for a tag key name. */
+  public static final int MAX_LENGTH = StringUtil.MAX_LENGTH;
```
I suggest making the max a much smaller number for initial purposes - maybe 255?

a-veitch commented on core/src/main/java/com/google/instrumentation/tags/TagKey.java:
```diff
@@ -0,0 +1,89 @@
+   * that:
+   *
+   * <ol>
+   *   <li>length is restricted to {@link #MAX_LENGTH}, strings longer than that will be truncated.
```
We should discuss what policies we want here. I think it would be better to raise an exception, rather than silently truncate or replace characters.

a-veitch commented on core/src/main/java/com/google/instrumentation/tags/TagKey.java:
```diff
@@ -0,0 +1,89 @@
+    return new AutoValue_TagKey<Boolean>(StringUtil.sanitize(key), TAG_BOOL);
+  }
+
+  abstract String getName();
```
Don't need either getName or getTagType in the v1 API.

a-veitch commented on core/src/main/java/com/google/instrumentation/tags/TagSet.java:
```diff
@@ -0,0 +1,188 @@
+ */
+public abstract class TagSet {
+  /** The maximum length for a string tag value. */
+  public static final int MAX_STRING_LENGTH = StringUtil.MAX_LENGTH;
```
Why is this here? Raises potential for conflict with the same constant in TagKey. Suggest having this in one place only (probably TagKey).

a-veitch commented on core/src/main/java/com/google/instrumentation/tags/TagSet.java:
```diff
@@ -0,0 +1,188 @@
+   * @param key the key to look up.
+   * @return {@code true} if the key is present.
+   */
+  public abstract boolean tagKeyExists(TagKey<?> key);
```
This implies that only one TagKey of a given name can exist at a time, which is not the case (you can have two tags with same key name, but different types). Need to match both name and type.

a-veitch commented on core/src/main/java/com/google/instrumentation/tags/TagSet.java:
```diff
@@ -0,0 +1,188 @@
+  public abstract boolean tagKeyExists(TagKey<?> key);
+
+  // TODO(sebright): Which is better, the generic "get" method or three specialized "get" methods?
+  // The specialized "get" methods can return primitives, which could be less expensive.
```
I think specialization is better, but don't feel strongly about it.

sebright replied:
Thanks.  I removed the generic "get" method.  The set/update/insert methods were already specialized.

a-veitch commented on core/src/main/java/com/google/instrumentation/tags/TagSet.java:
```diff
@@ -0,0 +1,188 @@
+  // TODO(sebright): Which is better, the generic "get" method or three specialized "get" methods?
+  // The specialized "get" methods can return primitives, which could be less expensive.
+
+  // TODO(sebright): Should these methods take defaults or throw exceptions to handle missing keys?
```
The trouble with default values is that there may not be one - especially for bool. I think that leaves your only other option as being throwing; if users are want to avoid that, they can use tagKeyExists first.

sebright replied:
I removed the default parameters.  I also made getStringTagValue throw an exception if the key is missing, for consistency.

a-veitch commented on core_impl/src/test/java/com/google/instrumentation/tags/TagSetTest.java:
```diff
@@ -0,0 +1,180 @@
+
+  private static final TagKey<String> KS1 = TagKey.createString("k1");
+  private static final TagKey<String> KS2 = TagKey.createString("k2");
+  private static final TagKey<Long> KI = TagKey.createInt("k2");
```
Should add a test doing something similar to (same key name, different types):

newBuilder().set(KS2, "v1").set(KI, 5).build()

sebright replied:
Good catch!

a-veitch commented on core_impl/src/main/java/com/google/instrumentation/internal/logging/Logger.java:
```diff
@@ -0,0 +1,28 @@
+/*
```
I understand after reading further - log on errors. I think we can do away with this for now, since we have defined error-less operations for v1, and even when we do have them, we will (probably) want to use exceptions instead.

sebright commented:
I rebased onto master and updated error handling for invalid tag key names.  Now the public TagKey factory methods throw IllegalArgumentException when the name is invalid.

sebright commented:
I temporarily removed all TagMap methods, except for `set` and `clear`, since it looks like we will definitely need those operations.  I can add other operations back in a separate PR, once we decide what we need.  This PR is ready for more review.

sebright commented on core/src/main/java/com/google/instrumentation/tags/TagMap.java:
```diff
@@ -0,0 +1,129 @@
+     */
+    public Builder set(TagKey<String> key, String value) {
+      Preconditions.checkArgument(key.getTagType() == TagType.TAG_STRING);
+      Preconditions.checkArgument(StringUtil.isValid(value));
```
Should we add a class to represent a string tag value, so that we can validate the string once and use it many times?

a-veitch replied:
That would be fine as an optimization. Up to you whether you want to do this now, or add a todo.

sebright replied:
Thanks.  I added a TODO.

a-veitch commented on core/src/main/java/com/google/instrumentation/tags/TagKey.java:
```diff
@@ -0,0 +1,126 @@
+   *
+   * <ol>
+   *   <li>It cannot be longer than {@link #MAX_LENGTH}.
+   *   <li>It can only contain printable ascii characters.
```
Nit: upper-case ASCII

sebright replied:
Done

a-veitch commented on core/src/main/java/com/google/instrumentation/tags/TagKey.java:
```diff
@@ -0,0 +1,126 @@
+    return createStringInternal(name);
+  }
+
+  // Creates a TagKey<String> by sanitizing the given String.
```
Not sure at all we want/need this functionality.

sebright replied:
I removed it.

a-veitch commented on core/src/main/java/com/google/instrumentation/tags/TagKey.java:
```diff
@@ -0,0 +1,126 @@
+   * @param name the name of the key.
+   * @throws IllegalArgumentException if the name is not valid.
+   */
+  public static TagKey<Long> createInt(String name) {
```
I think the name should be createLong as well - otherwise it's confusing.

a-veitch replied:
Also other names, e.g. s/TAG_INT/TAG_LONG/

sebright replied:
Done

a-veitch commented on core/src/main/java/com/google/instrumentation/tags/TagMap.java:
```diff
@@ -0,0 +1,129 @@
+    /**
+     * Adds the key/value pair regardless of whether the key is present.
+     *
+     * @param key the key to look up.
```
Suggest: @param key the {@code TagKey} which will be set

sebright replied:
Done

a-veitch commented on core/src/main/java/com/google/instrumentation/tags/TagMap.java:
```diff
@@ -0,0 +1,129 @@
+     * Adds the key/value pair regardless of whether the key is present.
+     *
+     * @param key the key to look up.
+     * @param value the value to set for the given key.
```
Need to add @throws documentation.

sebright replied:
Done

sebright commented:
I renamed the `tags` package to `io.opencensus.tags`.  The name isn't finalized, but it makes it clear that the package is part of the new API.

bogdandrutu commented on core/src/main/java/io/opencensus/tags/TagKey.java:
```diff
@@ -0,0 +1,111 @@
+  /** The maximum length for a tag key name. */
+  public static final int MAX_LENGTH = StringUtil.MAX_LENGTH;
+
+  enum TagType {
```
I thought we are going to support only String type initially.

sebright replied:
I made the factory methods for `TagKey<Long>` and `TagKey<Boolean>` non-public.

bogdandrutu commented on core/src/main/java/io/opencensus/tags/TagKey.java:
```diff
@@ -0,0 +1,111 @@
+    return createStringInternal(name);
+  }
+
+  private static TagKey<String> createStringInternal(String name) {
```
Why do you need this extra private method?

sebright replied:
Previously, there were two factory methods for each type of key, but the private methods aren't needed now.

bogdandrutu commented on core/src/main/java/io/opencensus/tags/TagKey.java:
```diff
@@ -0,0 +1,111 @@
+  TagKey() {}
+
+  /**
+   * Constructs a {@code TagKey<String>} from the given string. The string must meet the following
```
The comment is confusing "from the given string". Probably better to say something like "with the given {@code name}" or something similar.

Also I would move the second sentence on a different line:
<p>The string must meet the following requirements:

sebright replied:
Done

bogdandrutu commented on core/src/main/java/io/opencensus/tags/TagKey.java:
```diff
@@ -0,0 +1,111 @@
+   * @param name the name of the key.
+   * @throws IllegalArgumentException if the name is not valid.
+   */
+  public static TagKey<String> createString(String name) {
```
s/createString/createStringKey?

sebright replied:
Done.  I was trying to keep the name short, but createStringKey isn't much longer.

bogdandrutu commented on core/src/main/java/io/opencensus/tags/TagKey.java:
```diff
@@ -0,0 +1,111 @@
+ */
+@Immutable
+@AutoValue
+public abstract class TagKey<TagValueT> {
```
The current implementation defines equals/hashCode <name, type>. I think we haven't decided on this but you should be aware of this.

sebright replied:
Yes, it shouldn't matter while we are only allowing String tags.

bogdandrutu commented on core/src/test/java/io/opencensus/tags/TagKeyTest.java:
```diff
@@ -0,0 +1,67 @@
+
+  @Test
+  public void testGetName() {
+    assertThat(TagKey.createString("foo").getName()).isEqualTo("foo");
```
Maybe a test for getType?

sebright replied:
Done

bogdandrutu commented on core/src/main/java/io/opencensus/tags/TagKey.java:
```diff
@@ -0,0 +1,104 @@
+   * @throws IllegalArgumentException if the name is not valid.
+   */
+  public static TagKey<String> createStringKey(String name) {
+    Preconditions.checkArgument(StringUtil.isValid(name));
```
nit: Why not static import checkArguments? This is how we do everywhere else.

bogdandrutu commented on core/src/main/java/io/opencensus/tags/TagKey.java:
```diff
@@ -0,0 +1,104 @@
+  /**
+   * Constructs a {@code TagKey<String>} with the given name.
+   *
+   * <p>The string must meet the following requirements:
```
s/string/{@code name}

Probably everywhere else.

bogdandrutu commented on core/src/main/java/io/opencensus/tags/TagMap.java:
```diff
@@ -0,0 +1,137 @@
+     * @return a {@code TagMap} with the same tags as this builder.
+     */
+    public TagMap build() {
+      return new TagMap(new HashMap<TagKey<?>, Object>(tags));
```
collections.unmodifiablemap to avoid any mistake?

bogdandrutu commented on core/src/main/java/io/opencensus/tags/TagMap.java:
```diff
@@ -0,0 +1,137 @@
+    public Builder set(TagKey<String> key, String value) {
+      Preconditions.checkArgument(key.getTagType() == TagType.TAG_STRING);
+
+      // TODO(sebright): Consider adding a TagValue class to avoid validating the String every time
```
I think this is a good requirement to the design.

bogdandrutu commented on core/src/main/java/io/opencensus/tags/TagMap.java:
```diff
@@ -0,0 +1,137 @@
+     * @return this
+     * @throws IllegalArgumentException if the key is null or the key is the wrong type.
+     */
+    public Builder set(TagKey<Boolean> key, boolean value) {
```
not public please.

bogdandrutu commented on core/src/main/java/io/opencensus/tags/TagMap.java:
```diff
@@ -0,0 +1,137 @@
+     * @return this
+     * @throws IllegalArgumentException if the key is null or the key is the wrong type.
+     */
+    public Builder set(TagKey<Long> key, long value) {
```
not public please.

bogdandrutu commented on core/src/main/java/io/opencensus/tags/TagMap.java:
```diff
@@ -0,0 +1,137 @@
+  }
+
+  /** Builder for the {@link TagMap} class. */
+  public static final class Builder {
```
At one point I thought that we will support something like:

public static TagMap build(TagMap parent, List<TagMutation>);

I think this way we can avoid 1 of the 2 copies of the tags.

Maybe not in this PR because we want to move forward but might worth playing with that idea.

sebright replied:
@dinooliva said that users probably won't add the same tag values every time, so it's probably not worth adding the `TagMap build(TagMap parent, List)` method.

sebright commented:
Thanks!

adriancole commented on core/src/main/java/io/opencensus/tags/TagKey.java:
```diff
@@ -0,0 +1,104 @@
+   * @throws IllegalArgumentException if the name is not valid.
+   */
+  // TODO(sebright): Make this public once we support types other than String.
+  static TagKey<Long> createLongKey(String name) {
```
to save future pull request/issue comments, I'd put a comment (not necessarily javadoc) rationale on choosing long vs float or something else to represent numbers. There are good reasons to choose Long, but in the future when people ask, this will help others who may not have made the decision help explain why.

sebright replied:
I created an issue: #640

sebright commented on core/src/main/java/io/opencensus/tags/TagKey.java:
```diff
@@ -0,0 +1,104 @@
+   * @throws IllegalArgumentException if the name is not valid.
+   */
+  // TODO(sebright): Make this public once we support types other than String.
+  static TagKey<Long> createLongKey(String name) {
```
/cc @dinooliva
